nyxx_lavalink/lib/src/Cluster.dart

```diff
@@ -0,0 +1,256 @@
+
+    switch(map["cmd"]) {
+      case "DISPATCH":
+        this._eventDispatcher.dispatchEvent(map);
+        break;
+
+      case "LOG": {
+        Level? level;
+
+        switch (map["level"]) {
+          case "INFO":
+            level = Level.INFO;
+            break;
+
+          case "WARNING":
+            level = Level.WARNING;
+            break;
+        }
+
+        _logger.log(level!, map["message"]);
+      }
+      break;
+
+      case "EXITED": {
+        final nodeId = map["nodeId"]! as int;
+        this._nodes.remove(nodeId);
+        this._connectingNodes.remove(nodeId);
+
+        _logger.info("[Node $nodeId] Exited");
+      }
+      break;
+
+      case "CONNECTED": {
+        final node = this._connectingNodes.remove(map["nodeId"] as int);
+
+        if (node != null) {
+          this._nodes[node.options.nodeId] = node;
+
+          _logger.info("[Node ${map["nodeId"]}] Connected to lavalink");
+        }
+      }
+      break;
+
+      case "DISCONNECTED": {
+        final node = this._nodes.remove(map["nodeId"] as int);
+
+        if (node != null) {
+          this._connectingNodes[node.options.nodeId] = node;
+
+          // this makes possible for a player to be moved to another node
+          node._players.forEach((guildId, _) => this._nodeLocations.remove(guildId));
+
+          // Also delete the players, so them can be created again on another node
+          node._players.clear();
+
+          _logger.info("[Node ${map["nodeId"]}] Disconnected from lavalink");
+        }
+      }
+      break;
+    }
+  }
+
+  void _registerEvents() {
+    this._client.onVoiceServerUpdate.listen((event) async {
+      final node = this._nodes[this._nodeLocations[event.guild.id]];
+      if (node == null) {
+        return;
+      }
+
+      final player = node._players[event.guild.id];
+      if (player == null) {
+        return;
+      }
+
+      player._handleServerUpdate(event);
+    });
+
+    this._client.onVoiceStateUpdate.listen((event) async {
+      if (event.raw["d"]["user_id"] != _clientId.toString()) {
+        return;
+      }
+      if (event.state.guild == null) {
+        return;
+      }
+
+      final node = this._nodes[this._nodeLocations[event.state.guild!.id]];
+      if (node == null) {
+        return;
+      }
+
+      final player = node._players[event.state.guild!.id];
+      if (player == null) {
+        return;
+      }
+
+      player._handleStateUpdate(event);
+    });
+  }
+
+  /// Creates a new cluster ready to start adding connections
+  _Cluster(this._client, this._clientId, [Level? loggingLevel]) {
+    this._registerEvents();
+
+    this._eventDispatcher = _EventDispatcher(this);
+
+    this._receiveStream = this._receivePort.asBroadcastStream();
+
+    this._receiveStream.listen(_handleNodeMessage);
+
+    Logger.root.level = loggingLevel ?? Level.INFO;
+  }
+}
+
+/// Cluster of lavalink nodes
+class Cluster extends _Cluster {
```

Comment:
Why public `Cluster` extends private class, would be confusing to users i guess, and I dont see much use of that